src/Hazelcast.Net/CP/FencedLock.cs

```diff
@@ -326,17 +394,32 @@ public async Task UnlockAsync()
         /// <param name="sessionId"></param>
         /// <param name="releaseSession"></param>
         /// <exception cref="LockOwnershipLostException"></exception>
-        private void VerifyNoLockOnThread(long threadId, long sessionId, bool releaseSession = true)
+        private SemaphoreSlim VerifyNoLockOnThread(long threadId, long sessionId, bool releaseSession = true, bool force = false)
         {
-            if (_lockedThreadToSession.TryGetValue(threadId, out var lockedSessionId) && lockedSessionId != sessionId)
+            if (!_contextSemaphore.TryGetValue(threadId, out var semaphore))
```

Comment:
Going to add this as a single comment, so we can discuss: I think there's still a race condition here. If two tasks `TryGetValue` and both find no value, both will allocate a new semaphore and continue with that semaphore... this is not working, I am afraid.